build,ci,release: add Linux x86_64 FIPS build

rail opened the pull request:
* Added a script to build   [golang-fips](https://github.com/golang-fips/go). The resulted Go version has a `fips` suffix, so we can easily select it in Bazel.
* Added a new Bazel toolchain and build config used by the FIPS build only.
* The FIPS docker image has FIPS-compliant `openssl` installed.
* Added a script to run FIPS build in CI.
* Tweaked one of the regexes to be less greedy.
* Added the platform to the release process.

Epic: DEVINF-478
Fixes: DEVINF-634
Fixes: DEVINF-633

Release note (build change): As a part of this change we start
publishing a FIPS compliant tarball and docker image for the Linux
x86_64 platform. The build uses OpenSSL libraries for crypto operations
by `dlopen`ing the corresponding dynamic libraries.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96107)
<!-- Reviewable:end -->

rail commented:
This is just a draft that I'd like to get some feedback.

So far:
* I can build the toolchain. To simplify the process I used the same docker image RedHat uses to test there patches on.
* The new platform is not enabled in ci yet. Will need to poke around publish-artifacts and friends
* the rules_go will need some changes, maybe something [like this](https://gist.github.com/rail/593cd6a18d110e066f80e00e149fb950) (just hand waving for now).

rickystewart commented on pkg/release/build.go:
```diff
@@ -86,6 +90,8 @@ func TargetTripleFromPlatform(platform Platform) string {
 	switch platform {
 	case PlatformLinux:
```
Nit: can do `case PlatformLinux, PlatformLinuxFIPS:` instead of duplicating the arm.

rickystewart commented on build/teamcity/internal/release/build-and-publish-patched-go-fips.sh:
```diff
@@ -0,0 +1,48 @@
+
+set -euo pipefail
+
+google_credentials="$GOOGLE_EPHEMERAL_CREDENTIALS"
```
Nit: Would be good to have this be part of the same job that builds the rest of the `go` toolchains so we don't have to trigger multiple jobs to upgrade `go`.

rail replied:
Agree. I'll merge the scripts when I'm happy with the result.

rickystewart commented on pkg/release/build.go:
```diff
@@ -238,6 +244,8 @@ type Platform int
 const (
 	// PlatformLinux is the Linux x86_64 target.
 	PlatformLinux Platform = iota
+	// PlatformLinuxFIPS is the Linux x86_64 target built with FIPS support.
```
Nit: Do we want FIPS + ARM?

rail replied:
Yes. I just started with one platform for now, because I need to fix the header search issue when we cross build arm64 on amd64.

rail replied:
We decided to go with intel only for now.

rail committed:
build,ci,release: add Linux x86_64 FIPS build

* Added a script to build
  [golang-fips](https://github.com/golang-fips/go). The resulted Go
  version has a `fips` suffix, so we can easily select it in Bazel.
* Added a new Bazel toolchain and build config used by the FIPS build only.
* The FIPS docker image has FIPS-compliant `openssl` installed.
* Added a script to run FIPS build in CI.
* Tweaked one of the regexes to be less greedy.
* Added the platform to the release process.

Epic: DEVINF-478
Fixes: DEVINF-634
Fixes: DEVINF-633

Release note (build change): As a part of this change we start
publishing a FIPS compliant tarball and docker image for the Linux
x86_64 platform. The build uses OpenSSL libraries for crypto operations
by `dlopen`ing the corresponding dynamic libraries.

rail commented:
This version works as expected. The only annoying thing is that we need to edit `.bazelrc` to specify the version when we update the toolchain. I wish there was a way to "tag" toolchains and select them based on the tag.

rail commented:
If the PR looks good, I will need to:
* rebuild the Go toolchain, maybe using 1.19.4 for FIPS instead of current 1.19.5 (so we use the same version)
* Update the `download_go_sdk` entries for either both or just the FIPS entry.

rickystewart commented:
> I wish there was a way to "tag" toolchains and select them based on the tag.

This is *exactly* how Bazel's [toolchain selection](https://bazel.build/extending/toolchains) works.

rickystewart reviewed (commented):
Before we check this in can you add a check in CI to validate that the binaries produced do NOT have any FIPS stuff in them? My understanding is `go tool nm $BINARY | grep FIPS` should return no output if the binary is not FIPS-capable. The current behavior seems correct but it makes me a little anxious that `rules_go` is "accidentally" going to pick the FIPS toolchain when it shouldn't at some point. A check in `make-and-publish-build` to make sure the FIPS binary *does* actually have FIPS stuff would be good too.

Up to you whether we keep doing it this way or if you want to set up the toolchain the traditional way, but if it works, then it works. 🤷

rail commented:
> Before we check this in can you add a check in CI to validate that the binaries produced do NOT have any FIPS stuff in them? My understanding is `go tool nm $BINARY | grep FIPS` should return no output if the binary is not FIPS-capable. The current behavior seems correct but it makes me a little anxious that `rules_go` is "accidentally" going to pick the FIPS toolchain when it shouldn't at some point. A check in `make-and-publish-build` to make sure the FIPS binary _does_ actually have FIPS stuff would be good too.

Let me implement these before merging this change.

rail commented:
FTR. We chatted with Ricky and decided to take the current approach so we don't need to patch rules_go heavily. After I test the release process I will request a review for the resulted patch.

rail commented:
This is ready for  review. I cannot fully test it for the 23.1 release, because there is one blocking test failure, that blocks Make and Publish. But I tested it before the blocker, so it should just work.
Additionally I tested the 22.2 backport and it worked as expected.

rickystewart reviewed (changes requested):
Looks pretty good!

rail commented on build/teamcity/cockroach/ci/builds/build_linux_x86_64_fips.sh:
```diff
@@ -0,0 +1,12 @@
+source "$dir/teamcity-bazel-support.sh"  # For run_bazel
+
+tc_start_block "Run Bazel build"
+run_bazel build/teamcity/cockroach/ci/builds/build_impl.sh crosslinuxfips
```
Good point! Will do

rail commented:
Finally, passed all flakes and got a [dry-run release](https://teamcity.cockroachdb.com/buildConfiguration/Internal_Release_PublishCockroachRelease/8739254?buildTab=artifacts). 
<img width="653" alt="image" src="https://user-images.githubusercontent.com/42865/219692004-9f0a212e-45b8-46e0-9388-3db23e2729f5.png">

rickystewart commented on build/teamcity/cockroach/ci/builds/build_impl.sh:
```diff
@@ -27,8 +27,26 @@ then
+        fi
+    done
+fi
+if [[ $CONFIG == "crosslinux" ]]; then
```
Nice.

rail commented:
bors r=rickystewart
